grype/presenter/cyclonedxvex/document.go

```diff
@@ -0,0 +1,84 @@
+package cyclonedxvex
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/CycloneDX/cyclonedx-go"
+	"github.com/google/uuid"
+
+	"github.com/anchore/grype/grype/match"
+	"github.com/anchore/grype/grype/pkg"
+	"github.com/anchore/grype/grype/vulnerability"
+	"github.com/anchore/grype/internal"
+	"github.com/anchore/grype/internal/version"
+	"github.com/anchore/syft/syft/source"
+)
+
+// NewDocument returns a CycloneDX Document object populated with the SBOM and vulnerability findings.
+func NewDocument(packages []pkg.Package, matches match.Matches, srcMetadata *source.Metadata, provider vulnerability.MetadataProvider) (*cyclonedx.BOM, error) {
+	versionInfo := version.FromBuild()
+	doc := cyclonedx.NewBOM()
+	doc.SerialNumber = uuid.New().URN()
+	if srcMetadata != nil {
+		doc.Metadata = NewBomMetadata(internal.ApplicationName, versionInfo.Version, srcMetadata)
+	}
+
+	// attach matches
+	components := []cyclonedx.Component{}
+	vulnerabilities := []cyclonedx.Vulnerability{}
+
+	for _, p := range packages {
+		bomRef := p.PURL
+		if bomRef == "" {
+			bomRef = strings.Join([]string{string(p.Type), p.Name, p.Version}, "-")
+		}
+		externalRef := fmt.Sprintf("%s/1#%s", doc.SerialNumber, bomRef)
+		// make a new Component (by value)
+		component := cyclonedx.Component{
+			Type:       "library", // TODO: this is not accurate, syft does the same thing
+			Name:       p.Name,
+			Version:    p.Version,
+			PackageURL: p.PURL,
+			BOMRef:     bomRef,
+		}
+
+		var licenses cyclonedx.Licenses
+		for _, licenseName := range p.Licenses {
+			licenses = append(licenses, cyclonedx.LicenseChoice{
+				License: &cyclonedx.License{
+					Name: licenseName,
+				},
+			})
+		}
+		if len(licenses) > 0 {
+			// adding licenses to the Component
+			component.Licenses = &licenses
+		}
+
+		// mutate the Component
+
+		pkgMatches := matches.GetByPkgID(p.ID)
+
+		if len(pkgMatches) > 0 {
+			for _, m := range pkgMatches {
+				v, err := NewVulnerability(m, provider)
+				if err != nil {
+					return &cyclonedx.BOM{}, err
+				}
+				v.Affects = &[]cyclonedx.Affects{
```

Comment:
Question: should we add a default analysis element per vulnerability ? I'd assume you won't need to analyze any single vulnerability, but I might be wrong. Also, would it make sense to filter the vulnerability added to the VEX file per severity level e.g. severity > Medium ?  
```
      "analysis": {
        "state": "not_affected",        
        }
```